wait-online: add initial support for waiting for DNS

enr0n opened the pull request:
/cc @yuwata

Opening this as a draft to get initial comments on the general approach. Is this an acceptable option to add to `systemd-networkd-wait-online`? And, in the cases where `systemd-resolved` is active, is checking `sd_network_link_get_dns()` sufficient? 

**Proposal:**

Add a new flag, `--dns`, to systemd-networkd-wait-online to allow waiting for DNS to be configured. The `--dns` flag respects the `--ipv4` and `--ipv6` flags, as well as `--interface=` and `--any`.

When `systemd-resolved` is running, we check for DNS by checking the link state file (i.e. by calling `sd_network_link_get_dns()`). When `systemd-resolved` is not active, we fallback to parsing `/etc/resolv.conf`.

bluca commented:
I like the idea - do we need to bother with resolv.conf? For the other wait-online we only support networkd, and it's up to other implementations to provide a service that uses the wait-online target, we could just do the same here?

yuwata commented on src/network/wait-online/wait-online.c:
```diff
@@ -48,6 +49,7 @@ static int help(void) {
                "  -6 --ipv6                 Requires at least one IPv6 address\n"
                "     --any                  Wait until at least one of the interfaces is online\n"
                "     --timeout=SECS         Maximum time to wait for network connectivity\n"
+               "     --dns                  Requires at least one DNS server to be configured\n"
```
`to be configured` -> `accessible` or so?

yuwata commented on src/network/wait-online/manager.c:
```diff
@@ -133,6 +133,33 @@ static int manager_link_is_online(Manager *m, Link *l, const LinkOperationalStat
+                                log_link_debug_errno(l, r, "Invalid DNS server %s, ignoring: %m", *d);
+                        else
+                                found |= family == AF_INET ? ADDRESS_FAMILY_IPV4 :
+                                         (family == AF_INET6 ? ADDRESS_FAMILY_IPV6 : ADDRESS_FAMILY_NO);
```
I guess `()` is not necessary.

yuwata commented on src/network/wait-online/manager.c:
```diff
@@ -133,6 +133,33 @@ static int manager_link_is_online(Manager *m, Link *l, const LinkOperationalStat
+                        return log_link_debug_errno(l, SYNTHETIC_ERRNO(EADDRNOTAVAIL),
+                                                    "No DNS server is configured.");
+
+                if (needs_ipv4 && !(found & ADDRESS_FAMILY_IPV4))
```
Please use `FLAGS_SET()`

yuwata commented on src/network/wait-online/manager.c:
```diff
@@ -133,6 +133,33 @@ static int manager_link_is_online(Manager *m, Link *l, const LinkOperationalStat
+                        return log_link_debug_errno(l, SYNTHETIC_ERRNO(EADDRNOTAVAIL),
+                                                    "No IPv4 DNS server is configured.");
+
+                if (needs_ipv6 && !(found & ADDRESS_FAMILY_IPV6))
```
Also here.

yuwata commented on src/network/wait-online/link.h:
```diff
@@ -18,11 +18,13 @@ struct Link {
         unsigned flags;
 
         bool required_for_online;
+        bool requires_dns_for_online;
```
Unused?

yuwata reviewed (commented):
Please also check that a DNS server is accessible.
E.g. if DNS server 8.8.8.8 is configured, but the interface has only IPv4 link-local address, then we cannot access the server.

Also, please check if resolved already acquired the server. Even if networkd saves DNS servers to the state file, resolved may not read the file yet.

poettering commented on src/network/wait-online/link.c:
```diff
@@ -145,6 +152,90 @@ static int link_update_altnames(Link *l, sd_netlink_message *m) {
+
+        log_link_debug(l, "Attempting to read DNS servers from /etc/resolv.conf");
+
+        f = fopen("/etc/resolv.conf", "re");
```
i am really not a fan of this fallback path i must say, but maybe it's ok.

poettering commented on src/network/wait-online/link.c:
```diff
@@ -145,6 +152,90 @@ static int link_update_altnames(Link *l, sd_netlink_message *m) {
+
+        f = fopen("/etc/resolv.conf", "re");
+        if (!f)
+                return -errno;
```
please log about this at debug level, i.e. turn this line into `return log_debug_errno(errno, "…` with some vaguely useful log message, this is otherwise too hard to debug if there are issues anywhere

poettering commented on src/network/wait-online/link.c:
```diff
@@ -145,6 +152,90 @@ static int link_update_altnames(Link *l, sd_netlink_message *m) {
+
+                r = read_stripped_line(f, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return r;
```
similar here, debug log

poettering commented on src/network/wait-online/link.c:
```diff
@@ -145,6 +152,90 @@ static int link_update_altnames(Link *l, sd_netlink_message *m) {
+
+                        r = extract_first_word(&a, &word, NULL, 0);
+                        if (r < 0)
+                                return r;
```
and here (and everywhere else)

poettering commented on src/network/wait-online/link.c:
```diff
@@ -246,5 +337,9 @@ int link_update_monitor(Link *l) {
 
+        r = link_update_dns(l);
+        if (r < 0)
+                ret = log_link_debug_errno(l, r, "Failed to update DNS servers, ignoring: %m");
```
and then i'd drop this

poettering commented on src/network/wait-online/link.c:
```diff
@@ -145,6 +152,90 @@ static int link_update_altnames(Link *l, sd_netlink_message *m) {
+                if (r < 0)
+                        return r;
+
+                strv_free_and_replace(l->dns, dns);
```
hmm, i find this a bit sloppy tbh, i.e. running this fallback path once for each interface, and then claiming global DNS was actually local DNS is a bit sloppy, no?

I think we should just propagate an error here if we cannot connect to resolved, and store the inability to talk to resolved as part of the Link object. And then, when determining the online status, read the resolve.conf file once, associated to the Manager object if we see this issue.

enr0n commented:
> Hmm, the problem with this approach is that it is racy: afaics you are only subscribing to changes from networkd, but not resolved. Thus, it might happen that networkd gets a valid config, you then check resolved, but it hasn't processed the request yet, and you hence say "not online", and you'll never notice once resolved has caught up.

My understanding is that when resolved configures/modifies per-link DNS, this does end up recorded in networkd's link state, and hence an event is emitted. Testing locally seems to confirm this, too.

On the other hand, @yuwata did mention above that just because networkd's link state records the DNS server, that does not mean resolved has picked it up yet (e.g. I guess DNS servers issued via DHCP would be reflected in the networkd state immediately).

> I think the correct, race-free fix would be if we teach resolved's varlink interface a new call SubscribeDNS() which would be one of those methods that Varlink allows to have multiple replies. It's first reply would be the status qup, and then it would fire an event whenever the manager or per-link DNS server list changes (or search domain list). It would return a complete description of the DNS config, i.e. one object that contains subobjects for the global config and for each link config.
> 
> wait-online could then issue that, and simply wait until it gets a non-empty reply.
> 
> I think the patch for that wouldn't even be too involved at all.

That does sound like a more concrete approach this. And it's more generic/re-usable. I will work on something to that extent.

Thanks for reviewing!

enr0n commented:
> Please also check that a DNS server is accessible. E.g. if DNS server 8.8.8.8 is configured, but the interface has only IPv4 link-local address, then we cannot access the server.
> 
> Also, please check if resolved already acquired the server. Even if networkd saves DNS servers to the state file, resolved may not read the file yet.

Thanks for reviewing! Both of those points make sense, and I will address them in my next version.

poettering commented:
> My understanding is that when resolved configures/modifies per-link DNS, this does end up recorded in networkd's link state, and hence an event is emitted. Testing locally seems to confirm this, too.

It's the other way round: networkd learns DNS from DHCP/IPv6LL and such like and then resolved picks it up from there. But this is asynchronous (it's actually implemented via inotify). resolved might take a bit until it see networkd's changes.

poettering commented:
> Please also check that a DNS server is accessible.
> E.g. if DNS server 8.8.8.8 is configured, but the interface has only IPv4 link-local address, then we cannot access the server.

Note that this is a really tough problem btw, because routing might make surprising stuff connectible, that the network configuration otherwise suggests wouldn't be allowed. i.e. you could route 8.8.8.8 or some subnet covering it to some ipv4ll host and it would work. 

It might be OK'ish as an approximation would be to allocate an UDP socket, then issue connect() towards the DNS server address. This forces a routing decision without actually sending any traffic. This should fail if there are no routes to that host, but succeed if there are. And that's the info you need. You never actually send any traffic over that socket, but immediately close it.

enr0n commented:
I pushed a new version with the following changes:

- I dropped the `/etc/resolv.conf` fallback. I would rather add something later if there is a stronger need for it.
- I added `io.systemd.Resolve.Monitor.SubscribeDNSConfiguration` to make `resolved` the source of truth.
- I added logic to check if DNS servers are "accessible" on the interface, using the suggestion above.

A couple things I am not sure of yet:

- How/when should the global DNS servers be considered for an interface?
- Related to that, how should `FallbackDNS=` come in to play?

If we immediately check global DNS when a link does not its own DNS servers yet, _and_ fallback DNS are configured, then we would usually just immediately use fallback to determine if a link is configured. That doesn't feel quite right, but I am not sure.

yuwata commented on src/resolve/resolved-link.c:
```diff
@@ -1457,3 +1458,55 @@ bool link_negative_trust_anchor_lookup(Link *l, const char *name) {
+        LIST_FOREACH(servers, s, l->dns_servers) {
+                _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+
+                r = sd_json_variant_new_string(&v, dns_server_string_full(s));
+                if (r < 0)
+                        return r;
+
+                r = sd_json_variant_append_array(&dns, v);
```
sd_json_variant_append_arrayb() ?

enr0n replied:
Good idea, thanks.

yuwata commented on src/resolve/resolved-link.c:
```diff
@@ -1457,3 +1458,55 @@ bool link_negative_trust_anchor_lookup(Link *l, const char *name) {
+        LIST_FOREACH(domains, d, l->search_domains) {
+                _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+
+                r = sd_json_variant_new_string(&v, DNS_SEARCH_DOMAIN_NAME(d));
+                if (r < 0)
+                        return r;
+
+                r = sd_json_variant_append_array(&search_domains, v);
```
Also here?

yuwata commented on src/resolve/resolved-link.c:
```diff
@@ -1457,3 +1458,55 @@ bool link_negative_trust_anchor_lookup(Link *l, const char *name) {
         return false;
 }
+
+int link_dump_configuration_json(Link *l, sd_json_variant **ret) {
```
link_dump_configuration_json() and manager_dump_...() has mostly same logic. Maybe, better to extract the logic?

enr0n replied:
Good point. I have consolidated this into a helper function in `resolved-manager.c` instead.

yuwata commented on src/resolve/resolved-manager.c:
```diff
@@ -1929,3 +1930,99 @@ void dns_manager_reset_statistics(Manager *m) {
+
+        r = manager_dump_dns_configuration_json(m, &configuration);
+        if (r < 0)
+                return log_error_errno(r, "Failed to dump DNS configuration json: %m");
+
+        r = varlink_many_notify(m->varlink_dns_configuration_subscription, configuration);
+        if (r < 0)
+                log_error_errno(r, "Failed to send DNS configuration event, ignoring: %m");
```
Please downgrade the log level to warning.
Also, please propagate the error, even though it is ignored by the caller.

yuwata commented on src/network/wait-online/link.c:
```diff
@@ -248,3 +256,47 @@ int link_update_monitor(Link *l) {
+                union in_addr_union in_addr;
+                int family;
+
+                r = in_addr_from_string_auto(*d, &family, &in_addr);
```
The string may contain port number and interface specifier.
Please use `in_addr_port_ifindex_name_from_string()`.

yuwata commented on src/resolve/resolved-manager.c:
```diff
@@ -1929,3 +1930,112 @@ void dns_manager_reset_statistics(Manager *m) {
+
+        r = manager_dump_dns_configuration_json(m, &configuration);
+        if (r < 0)
+                return log_error_errno(r, "Failed to dump DNS configuration json: %m");
```
Please also downgrade this to LOG_WARNING.

yuwata commented on src/resolve/resolved-manager.c:
```diff
@@ -1929,3 +1930,112 @@ void dns_manager_reset_statistics(Manager *m) {
+
+        r = varlink_many_notify(m->varlink_dns_configuration_subscription, configuration);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to send DNS configuration event, ignoring: %m");
```
When an error is propagated, please drop ", ignoring".

yuwata commented on src/resolve/resolved-varlink.c:
```diff
@@ -125,14 +125,21 @@ static void vl_on_disconnect(sd_varlink_server *s, sd_varlink *link, void *userd
                 sd_varlink_unref(removed_link);
-                log_debug("%u monitor clients remain active", set_size(m->varlink_query_results_subscription));
+                log_debug("%u query result monitor clients remain active", set_size(m->varlink_query_results_subscription));
+        }
```
Not a matter, but I'd like to move these changes for query result monitor to the earlier commit.

yuwata commented on src/resolve/resolved-varlink.c:
```diff
@@ -1280,7 +1287,7 @@ static int vl_method_subscribe_query_results(sd_varlink *link, sd_json_variant *
         sd_varlink_ref(link);
 
-        log_debug("%u clients now attached for varlink notifications", set_size(m->varlink_query_results_subscription));
+        log_debug("%u clients now attached for query result varlink notifications", set_size(m->varlink_query_results_subscription));
```
Also this one. Please move to the earlier commit.

yuwata commented on src/resolve/resolved-manager.c:
```diff
@@ -1929,3 +1930,112 @@ void dns_manager_reset_statistics(Manager *m) {
+                DnsServer *current_dns_server,
+                DnsServer *dns_servers,
+                DnsSearchDomain *search_domains,
+                sd_json_variant **ret) {
```
Not a matter, but by taking the array rather than 'ret', you can use `sd_json_variant_appendbo()`.

yuwata commented on src/resolve/resolved-manager.c:
```diff
@@ -1929,3 +1930,112 @@ void dns_manager_reset_statistics(Manager *m) {
+        if (r < 0)
+                return r;
+
+        r = sd_json_variant_append_array(&configuration, global);
```
Then, _append_arra() here and below can be dropped.

yuwata commented on src/resolve/resolved-manager.c:
```diff
@@ -1929,3 +1930,112 @@ void dns_manager_reset_statistics(Manager *m) {
+                return r;
+
+        LIST_FOREACH(servers, s, dns_servers) {
+                r = sd_json_variant_append_arrayb(&dns_servers_json, SD_JSON_BUILD_STRING(dns_server_string_full(s)));
```
By reconsidering about this, I think it is better to build structured information for DNS server. That is, store address (byte array), port (unsigned), interface specifier (integer), server name (string). Like resolved already does in DBus and other varlink methods.

poettering replied:
yes, please generate a proper json representation of the server fields.

enr0n replied:
Thanks, I have done this in the latest push.

poettering commented on src/resolve/resolved-manager.c:
```diff
@@ -1929,3 +1930,112 @@ void dns_manager_reset_statistics(Manager *m) {
+        }
+
+        LIST_FOREACH(domains, d, search_domains) {
+                r = sd_json_variant_append_arrayb(&search_domains_json, SD_JSON_BUILD_STRING(DNS_SEARCH_DOMAIN_NAME(d)));
```
and similar here.

poettering commented on src/resolve/resolved-manager.c:
```diff
@@ -1929,3 +1930,112 @@ void dns_manager_reset_statistics(Manager *m) {
+
+        return sd_json_buildo(
+                        ret,
+                        SD_JSON_BUILD_PAIR_STRING("interface", ifname),
```
use JSON_BUILD_PAIR_STRING_NON_EMPTY() to suppress the field if no interface is specified

poettering commented on src/resolve/resolved-manager.c:
```diff
@@ -1929,3 +1930,112 @@ void dns_manager_reset_statistics(Manager *m) {
+        return sd_json_buildo(
+                        ret,
+                        SD_JSON_BUILD_PAIR_STRING("interface", ifname),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("interfaceIndex", ifindex),
```
use SD_JSON_BUILD_PAIR_CONDITION() here to ensure the field is suppressed if ifindex is <= 0

poettering commented on src/resolve/resolved-manager.c:
```diff
@@ -1929,3 +1930,112 @@ void dns_manager_reset_statistics(Manager *m) {
+                        SD_JSON_BUILD_PAIR_STRING("interface", ifname),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("interfaceIndex", ifindex),
+                        SD_JSON_BUILD_PAIR_STRING("currentDNSServer",
+                                                  current_dns_server ? dns_server_string_full(current_dns_server) : NULL),
```
this here should use SD_JSON_BUILD_PAIR_CONDITION to suppress the field entirely if there is no current DNS server

poettering commented on src/network/wait-online/wait-online.c:
```diff
@@ -48,6 +49,7 @@ static int help(void) {
                "  -6 --ipv6                 Requires at least one IPv6 address\n"
                "     --any                  Wait until at least one of the interfaces is online\n"
                "     --timeout=SECS         Maximum time to wait for network connectivity\n"
+               "     --dns                  Requires at least one DNS server to be accessible\n"
```
I'd prefer if this was an "extended boolean" enum. i.e. that parses as boolean, but also can take an additional value for disabling the connectivity check (because I think it's problematic in some environments, because resolved might be allow to talk to DNS while other stuff might not be). Maybe call that extra value "passive" or so?

i.e. `--dns=yes`, `--dns=no`, `--dns=passive`

this is easy to do via DEFINE_PRIVATE_STRING_TABLE_LOOKUP_FROM_STRING_WITH_BOOLEAN() and related boilerplate macros

enr0n replied:
I guess this is no longer needed when doing the check from resolved, right?

poettering replied:
dunno, it'd at least make it take a boolean for now, so that we can extend it later.

i.e. `--dns=yes`. we have a simpler parser for that parse_boolean_argument()

poettering commented on src/network/wait-online/link.c:
```diff
@@ -248,3 +257,54 @@ int link_update_monitor(Link *l) {
+                else
+                        accessible |= family == AF_INET ? ADDRESS_FAMILY_IPV4 :
+                                      family == AF_INET6 ? ADDRESS_FAMILY_IPV6 : ADDRESS_FAMILY_NO;
+        }
```
i think it woul be wise to exit the loop once `accessible` has both families set.

poettering commented on src/network/wait-online/link.c:
```diff
@@ -248,3 +257,54 @@ int link_update_monitor(Link *l) {
+                        continue;
+                }
+
+                fd = socket(family, SOCK_DGRAM, 0);
```
i our codebase we want that O_CLOEXEC is set for *all* fds by default. i.e. please add SOCK_CLOEXEC here

poettering commented on src/network/wait-online/manager.c:
```diff
@@ -381,13 +406,122 @@ static int manager_network_monitor_listen(Manager *m) {
+                        continue;
+                }
+
+                w = sd_json_variant_by_key(v, "interfaceIndex");
```
these days we prefer if sd_json_dispatch() is used to map json fields to a C struct

enr0n replied:
Thanks, I have revised this.

enr0n commented:
> Hmm, thinking about this, I think the reachability check should probably be moved from the client side to the server side, i.e. into resolved. And then be reported in an additional boolean json field. That makes sure we always use the networking context of the server for all reachability checks, which is a really good thing.

This makes sense. I will change that.

> there should be one of these fields per link, and one for the global scope. The global scope one should use fallback servers if there are no configured ones, under the same rules as usual. see manager_get_dns_server().

Shouldn't there be one field per DNS server? That way we know which specific servers are accessible?

poettering commented:
> Shouldn't there be one field per DNS server? That way we know which specific servers are accessible?

Well, we'd only do a routability check, not an actual communication check. I wonder if such fine grained info is useful if it doesn't actually check if the DNS server is really responding... The only reason I asked for the per-link booleans is that each link might have different routing domains configured, and hence it is very important to know the difference between "is the DNS server where we send all kinds of lookups to accessible?" and "is the DNS server used for this exotic subdomain only accessible?".

Anyway, up to you. if you want to make it even more detailed, fine by me, but I personally think just one boolean per "scope" is enough for us to communicate.

poettering commented:
(sorry for the close, i misclicked)

enr0n commented:
I will look at all the failing CI tomorrow.

I also still need to add logic for using the global and/or fallback DNS servers to satisfy the wait-online state.

enr0n commented:
In the last push, I changed the logic for when wait-online considers DNS configured for a link.

When a link has search domains configured, or it is `DNSDefaultRoute=yes`, then we insist that link-specific DNS servers are configured. Otherwise, we allow global DNS servers to satisfy the condition.

I am trying to be consistent with what resolved does here according to [the man page](https://www.freedesktop.org/software/systemd/man/latest/systemd-resolved.service.html#Protocols%20and%20Routing).

enr0n commented:
I looked at the remaining CI failures, and I believe they are unrelated. PTAL.

poettering commented on src/resolve/resolved-link.c:
```diff
@@ -299,6 +299,12 @@ static int link_update_dns_servers(Link *l) {
+        LIST_FOREACH(servers, s, l->dns_servers)
+                /* If the link state changed, we should re-check if DNS servers
+                 * are accessible. */
+                s->accessible = -1;
```
can you add a helper for that?

dns_server_reset_accessible()?

poettering commented on src/resolve/resolved-dns-server.c:
```diff
@@ -1164,3 +1166,28 @@ int dns_server_is_accessible(DnsServer *server) {
+        return sd_json_buildo(
+                        ret,
+                        JSON_BUILD_PAIR_IN_ADDR("address", &server->address, server->family),
+                        SD_JSON_BUILD_PAIR_INTEGER("addressFamily", server->family),
```
we typically call address family fields just "family".

poettering commented on src/resolve/resolved-dns-server.c:
```diff
@@ -1164,3 +1166,28 @@ int dns_server_is_accessible(DnsServer *server) {
+                        JSON_BUILD_PAIR_IN_ADDR("address", &server->address, server->family),
+                        SD_JSON_BUILD_PAIR_INTEGER("addressFamily", server->family),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("port", dns_server_port(server)),
+                        SD_JSON_BUILD_PAIR_CONDITION(ifindex > 0, "interfaceSpecifier", SD_JSON_BUILD_UNSIGNED(ifindex)),
```
and ifindex fields we call just "ifindex"

poettering commented on src/resolve/resolved-dns-server.c:
```diff
@@ -1164,3 +1166,28 @@ int dns_server_is_accessible(DnsServer *server) {
+                        SD_JSON_BUILD_PAIR_INTEGER("addressFamily", server->family),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("port", dns_server_port(server)),
+                        SD_JSON_BUILD_PAIR_CONDITION(ifindex > 0, "interfaceSpecifier", SD_JSON_BUILD_UNSIGNED(ifindex)),
+                        JSON_BUILD_PAIR_STRING_NON_EMPTY("serverName", server->server_name),
```
i'd just call this "name"

poettering commented on src/resolve/resolved-manager.c:
```diff
@@ -1932,3 +1934,114 @@ void dns_manager_reset_statistics(Manager *m) {
+
+        return sd_json_variant_append_arraybo(
+                        configuration,
+                        JSON_BUILD_PAIR_STRING_NON_EMPTY("interface", ifname),
```
call this ifname

poettering commented on src/resolve/resolved-manager.c:
```diff
@@ -1932,3 +1934,114 @@ void dns_manager_reset_statistics(Manager *m) {
+        return sd_json_variant_append_arraybo(
+                        configuration,
+                        JSON_BUILD_PAIR_STRING_NON_EMPTY("interface", ifname),
+                        SD_JSON_BUILD_PAIR_CONDITION(ifindex > 0, "interfaceIndex", SD_JSON_BUILD_UNSIGNED(ifindex)),
```
and this ifindex

poettering commented on src/resolve/resolved-manager.c:
```diff
@@ -1932,3 +1934,114 @@ void dns_manager_reset_statistics(Manager *m) {
+                        JSON_BUILD_PAIR_STRING_NON_EMPTY("interface", ifname),
+                        SD_JSON_BUILD_PAIR_CONDITION(ifindex > 0, "interfaceIndex", SD_JSON_BUILD_UNSIGNED(ifindex)),
+                        JSON_BUILD_PAIR_VARIANT_NON_NULL("currentDNSServer", current_dns_server_json),
+                        SD_JSON_BUILD_PAIR_VARIANT("dnsServers", dns_servers_json),
```
"currentDNSServer" uses uppercase "DNS", but "dnsServers" uses lower-case. Not sure what is better. maybe just say "currentServer" and "servers"?

poettering commented on src/shared/varlink-io.systemd.Resolve.Monitor.c:
```diff
@@ -112,6 +112,30 @@ static SD_VARLINK_DEFINE_METHOD(
+                SD_VARLINK_DEFINE_FIELD(port, SD_VARLINK_INT, 0),
+                SD_VARLINK_DEFINE_FIELD(interfaceSpecifier, SD_VARLINK_INT, SD_VARLINK_NULLABLE),
+                SD_VARLINK_DEFINE_FIELD(serverName, SD_VARLINK_STRING, SD_VARLINK_NULLABLE),
+                SD_VARLINK_DEFINE_FIELD(accessible, SD_VARLINK_BOOL, 0));
```
for new varlink IDL please always add full comments, i.e. use SD_VARLINK_FIELD_COMMENT(), SD_VARLINK_INTERFACE_COMMENT(), SD_VARLINK_SYMBOL_COMMENT().

poettering commented on src/libsystemd/sd-json/json-util.c:
```diff
@@ -124,6 +124,22 @@ int json_dispatch_in_addr(const char *name, sd_json_variant *variant, sd_json_di
         return 0;
 }
 
+int json_dispatch_in_addr_union(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata) {
```
hmm, i am not sure how i feal about this one. 

i.e. the size of this should really verified whether it matches the expected address size of the family. But that's a check that only can be done once the full set of parameters is processed.

This could be used for bad stuff otherwise, for example in this current version if you pass only 4 bytes of address but pick ipv6 as address format this would happily accept that and then later run into uninitialized memory...

enr0n replied:
I took the suggestion to use the iovec dispatcher instead.

poettering commented on src/network/wait-online/dns-configuration.c:
```diff
@@ -0,0 +1,131 @@
+
+static int dispatch_dns_server(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata) {
+        static const sd_json_dispatch_field dns_server_dispatch_table[] = {
+                { "address",            SD_JSON_VARIANT_ARRAY,    json_dispatch_in_addr_union, offsetof(DNSServer, addr),        SD_JSON_MANDATORY },
```
i'd just use the iovec dispatcher here i guess. and then convert it into an address further down after verifying the size

poettering commented on src/network/wait-online/dns-configuration.c:
```diff
@@ -0,0 +1,131 @@
+
+static int dispatch_dns_server_array(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata) {
+        Set **ret = ASSERT_PTR(userdata);
+        Set *dns_servers = set_new(NULL);
```
missing oom ceck

poettering commented on src/network/wait-online/dns-configuration.c:
```diff
@@ -0,0 +1,131 @@
+                return NULL;
+
+        dns_server_free(c->current_dns_server);
+        set_free_with_destructor(c->dns_servers, dns_server_free);
```
please do not use set_free_with_destructor() anymore. it only exists for legacy code. in new code please allocate the set with a hash_ops that has the right destructor hooked in. that way code can just call set_free() without knowing anything .

use DEFINE_HASH_OPS_WITH_VALUE_DESTRUCTOR() macro to define this

poettering commented on src/network/wait-online/dns-configuration.c:
```diff
@@ -0,0 +1,131 @@
+static int dispatch_dns_configuration(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata) {
+        static const sd_json_dispatch_field dns_configuration_dispatch_table[] = {
+                { "interface",        SD_JSON_VARIANT_STRING,   sd_json_dispatch_string,   offsetof(DNSConfiguration, ifname),             0 },
+                { "interfaceIndex",   SD_JSON_VARIANT_UNSIGNED, sd_json_dispatch_uint,     offsetof(DNSConfiguration, ifindex),            0 },
```
hmm, we already have json_dispatch_ifindex() in nss-resolve, we should generalize that and refuse that here, it sounds generic enough to be useful at various places.

poettering replied:
→ #34982

poettering commented on src/resolve/resolved-manager.c:
```diff
@@ -1932,3 +1934,114 @@ void dns_manager_reset_statistics(Manager *m) {
+                                link_get_dns_server(l),
+                                l->dns_servers,
+                                l->search_domains,
+                                &configuration);
```
hmm, `l->default_route` should be serialized here too

poettering commented on src/resolve/resolved-manager.c:
```diff
@@ -1932,3 +1934,114 @@ void dns_manager_reset_statistics(Manager *m) {
+        }
+
+        LIST_FOREACH(domains, d, search_domains) {
+                r = sd_json_variant_append_arrayb(&search_domains_json, SD_JSON_BUILD_STRING(DNS_SEARCH_DOMAIN_NAME(d)));
```
i think these should be structs as well, so that we can properly serialize the "route_only" flag of each search domain

enr0n replied:
Done in the latest push.

poettering commented on src/network/wait-online/link.c:
```diff
@@ -246,5 +246,11 @@ int link_update_monitor(Link *l) {
         else
                 free_and_replace(l->state, state);
 
+        r = sd_network_link_get_dns_default_route(l->ifindex);
```
this information should be provided by resolved I am sure. (see comment on earlier commit)

poettering commented on man/systemd-networkd-wait-online.service.xml:
```diff
@@ -144,6 +144,20 @@
+        becomes online and has an accessible DNS server.</para>
+        <para>If there are search domains configured on a link, or if a link has <varname>DNSDefaultRoute=yes</varname>,
+        then there must be an accessible link-specific DNS server configured. Otherwise, global DNS servers
+        may be considered.</para>
```
hmm, i think we should just look for interfaces where default route is on, or where the "." domain is listed as search domain, and ignore all interfaces whether default route is off and no "." search domain is listed.

enr0n replied:
Can you elaborate a bit on your reasoning here? Why only consider the '.' domain? I was thinking that if any domains are configured on a link, then servers also need to be configured for those domains to be useful, no?

enr0n replied:
Thanks for clarifying. I think the latest push addresses this better.

poettering commented on src/network/wait-online/manager.c:
```diff
@@ -133,6 +139,28 @@ static int manager_link_is_online(Manager *m, Link *l, const LinkOperationalStat
+                if (l->dns_configuration)
+                        require_link_dns = !strv_isempty(l->dns_configuration->search_domains);
+
+                require_link_dns = require_link_dns || l->dns_default_route;
```
as mentioned above I think this check should be:

```c
have_dns = strv_contains(search_domains, ".") || l->dns_default_route;
```

poettering commented:
looks pretty good, i like where this is going

enr0n commented:
@poettering gentle ping

bluca reviewed (approved):
This is a very nice feature, and it looks ready to me? If there are no further comments we can merge it

yuwata commented:
Sorry for late, but I will re-review later.

yuwata commented on test/networkd-test.py:
```diff
@@ -1095,6 +1095,134 @@ def get_tz():
             self.show_journal('systemd-timedated.service')
             self.fail(f'Timezone: {tz}, expected: Pacific/Honolulu')
 
+    def do_test_wait_online_dns(
```
Not bad, but nowadays new test cases are added in test/test-network/systemd-networkd-tests.py . Could you move the test case into it?
Otherwise, the test case is only tested on Ubuntu CIs, while the new test runs on many distro CIs.

yuwata reviewed (changes requested):
Looks super nice. But several memleak and so on.

poettering commented on src/resolve/resolved-dns-server.c:
```diff
@@ -1128,3 +1134,30 @@ int dns_server_dump_state_to_json(DnsServer *server, sd_json_variant **ret) {
+                return -errno;
+
+        r = RET_NERRNO(connect(fd, &sa.sa, SOCKADDR_LEN(sa)));
+        if (!IN_SET(r, 0, -ENETUNREACH))
```
this should probably list mor errors, such as:

```c
                      -EHOSTDOWN,
                      -EHOSTUNREACH,
                      -ENETDOWN,
                      -ENETRESET,
                      -ENONET,
```

poettering commented on src/resolve/resolved-manager.c:
```diff
@@ -196,6 +197,40 @@ static int manager_process_address(sd_netlink *rtnl, sd_netlink_message *mm, voi
+        }
+
+        LIST_FOREACH(servers, s, m->dns_servers)
+                dns_server_reset_accessible(s);
```
can you introduce a new helper call manager_reset_all_servers_accessible() or so that just cntains these two loops?

poettering commented on src/resolve/resolved-manager.c:
```diff
@@ -196,6 +197,40 @@ static int manager_process_address(sd_netlink *rtnl, sd_netlink_message *mm, voi
+                dns_server_reset_accessible(s);
+
+        (void) manager_send_dns_configuration_changed(m);
+
```
so i am a bit concerned that we'll run through this for a myriad of routes (because plenty of systems have a *lot* of routes, and expect resolved to deal with this). if the system also has a lot of interfaces this is a lot of work, and that over and over again. and this might send out IPC notifications, which is even worse

one idea: move this into a defer job. i.e. and sd_event_add_defer() job that runs at a low priority and is enabled whenever we see RTM_NEWROUTE/RTM_DELROUTE. that defer job would then do the actual work, when the event loop is idle again

poettering replied:
also, maybe extract the ifindex from the route msg (if there is one), and only invalidate the accessible flag of the relevant interfaces (and of the global servers).

just to reduce impact

poettering replied:
another idea: maybe only enable the RTM_NEWROUTE/RTM_DELROUTE matches while at least one client cares? and disable them when the last one is gone? 

that would substantially reduce the chance that people start complaining about this

enr0n replied:
I added logic to only enable this match when there is at least one client connected. I also added logic to only emit the notification when the resulting JSON has changed, in an attempt to reduce IPC notifications related to this.

enr0n commented on src/network/wait-online/dns-configuration.c:
```diff
@@ -0,0 +1,244 @@
+                return r;
+
+        /* We ignored the address field above. Handle it now. */
+        r = json_dispatch_byte_array_iovec("address", sd_json_variant_by_key(variant, "address"), flags, &iov);
```
Mostly because I wanted to avoid adding the extra struct field. We need the extra bit of logic to verify the size anyways, so I thought I would keep it all contained in `dispatch_dns_server`.

I can add an `addr_iov` field or so to `DNSServer` if you think that's better.

enr0n commented on src/resolve/resolved-link.c:
```diff
@@ -299,6 +299,15 @@ static int link_update_dns_servers(Link *l) {
+                dns_server_reset_accessible(s);
+
+        LIST_FOREACH(servers, s, l->manager->dns_servers)
+                dns_server_reset_accessible(s);
```
I added `manager_send_configuration_changed()` calls to the other netlink match handlers.

yuwata commented on src/resolve/resolved-dns-server.c:
```diff
@@ -1128,3 +1134,50 @@ int dns_server_dump_state_to_json(DnsServer *server, sd_json_variant **ret) {
+                    -ENETDOWN,
+                    -ENETRESET,
+                    -ENONET))
+
```
Please drop the empty line.

yuwata commented on src/resolve/resolved-link.c:
```diff
@@ -299,6 +299,9 @@ static int link_update_dns_servers(Link *l) {
 
         dns_server_unlink_marked(l->dns_servers);
+
+        dns_server_reset_accessible_all(l->dns_servers);
```
Maybe also reset in bus_link_method_set_dns_servers_internal() and bus_link_method_revert() ?

enr0n commented:
@yuwata Fixed and rebased.

yuwata commented on src/network/wait-online/manager.c:
```diff
@@ -381,13 +407,102 @@ static int manager_network_monitor_listen(Manager *m) {
+
+                if (c->ifindex > 0) {
+                        Link *l = hashmap_get(m->links_by_index, INT_TO_PTR(c->ifindex));
+                        if (l)
```
If we do not know the interface at this point, the acquired data should be managed by `Manager` object. Otherwise, if we receiver `RTM_NEWLINK` soon after this event, we lost the DNS information for the interface, and may delay the event loop.
`link_new()` should find the DNS info in the `Manager` then.

yuwata commented on units/systemd-networkd-wait-online.service.in:
```diff
@@ -14,7 +14,7 @@ ConditionCapability=CAP_NET_ADMIN
 Conflicts=shutdown.target
 BindsTo=systemd-networkd.service
-After=systemd-networkd.service
+After=systemd-networkd.service systemd-resolved.service
```
Please drop the change. `--dns` is not enabled in the command line, hence not necessary.

yuwata commented on src/network/wait-online/manager.c:
```diff
@@ -381,13 +407,102 @@ static int manager_network_monitor_listen(Manager *m) {
+
+        assert(link);
+
+        if (error_id) {
```
I think we should reconnect the server when disconnected, for the case that resolved is restarted.

Or, maybe better that resolved serializes connections on restart.

enr0n replied:
Serializing connections on restart seems like the better solution to me. Could I handle that in a separate follow-up PR?

yuwata replied:
Yeah. I think it's ok to implement that later.

evverx replied:
FWIW it also gives up immediately if resolved is down for whatever reason so it doesn't actually take the timeout into account in all cases.

yuwata commented on src/resolve/resolved-manager.c:
```diff
@@ -108,6 +109,11 @@ static int manager_process_link(sd_netlink *rtnl, sd_netlink_message *mm, void *
         manager_mdns_maybe_stop(m);
+
+        /* The accessible flag on link DNS servers will have been reset by
+         * link_update. Just reset the global DNS servers. */
```
Not a matter, but please suffix the function name with `()`.

yuwata commented on src/resolve/resolved-manager.c:
```diff
@@ -192,13 +198,46 @@ static int manager_process_address(sd_netlink *rtnl, sd_netlink_message *mm, voi
+
+        r = sd_netlink_message_read_u32(mm, RTA_OIF, &ifindex);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to get route ifindex, ignoring: %m");
```
Some routes do not have RTA_OIF. Please downgrade the log level to debug when `-ENODATA`.

enr0n replied:
I guess we still need to call `manager_send_dns_configuration_changed()` in such cases. I will re-work that.

yuwata replied:
Ah, yes. Indeed.

yuwata reviewed (changes requested):
Looks much much better. But several further comments.

enr0n commented:
@yuwata Besides the resolved serialization, I believe I addressed all other comments.

yuwata reviewed (approved):
One more request. LGTM except for that.

enr0n commented on src/network/wait-online/manager.c:
```diff
@@ -381,13 +407,114 @@ static int manager_network_monitor_listen(Manager *m) {
+                                free_and_replace_full(l->dns_configuration, c, dns_configuration_free);
+                        else {
+                                r = hashmap_ensure_put(
+                                                &m->dns_configuration_by_link_index,
```
Done :+1:

yuwata reviewed (approved):
LGTM. Thank you!

bluca commented:
This looks related @enr0n can you double check?

```
[   98.549876] systemd-networkd-tests.py[691]:   File "/usr/lib64/python3.13/subprocess.py", line 508, in check_returncode
[   98.551901] systemd-networkd-tests.py[691]:     raise CalledProcessError(self.returncode, self.args, self.stdout,
[   98.5529[45](https://github.com/systemd/systemd/actions/runs/13002269118/job/36263024103?pr=34640#step:12:46)] systemd-networkd-tests.py[691]:                              self.stderr)
[   98.553440] systemd-networkd-tests.py[691]: subprocess.CalledProcessError: Command '['/usr/lib/systemd/systemd-networkd-wait-online', '--timeout=20s', '--interface=veth99:routable', '--ignore=sit0', '--ignore=ifb1', '--ignore=gre0', '--ignore=ifb0', '--ignore=lo', '--ignore=ip6tnl0', '--ignore=ip_vti0', '--ignore=ip6_vti0', '--ignore=teql0', '--ignore=ip6gre0', '--ignore=tunl0', '--ignore=gretap0', '--ignore=erspan0', '--dns']' returned non-zero exit status 1.
[   98.553822] systemd-networkd-tests.py[691]: During handling of the above exception, another exception occurred:
[   98.554372] systemd-networkd-tests.py[691]: Traceback (most recent call last):
[   98.554753] systemd-networkd-tests.py[691]:   File "/usr/lib/systemd/tests/testdata/test-network/systemd-networkd-tests.py", line 1696, in test_wait_online_dns_global
[   98.555136] systemd-networkd-tests.py[691]:     self.do_test_wait_online_dns(
[   98.555498] systemd-networkd-tests.py[691]:     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
[   98.556420] systemd-networkd-tests.py[691]:         global_dns='192.168.5.1',
[   98.557201] systemd-networkd-tests.py[691]:         ^^^^^^^^^^^^^^^^^^^^^^^^^
[   98.557664] systemd-networkd-tests.py[691]:     ...<3 lines>...
[   98.558110] systemd-networkd-tests.py[691]:         )
[   98.558487] systemd-networkd-tests.py[691]:         ^
[   98.558915] systemd-networkd-tests.py[691]:     )
[   98.560310] systemd-networkd-tests.py[691]:     ^
[   98.560550] systemd-networkd-tests.py[691]:   File "/usr/lib/systemd/tests/testdata/test-network/systemd-networkd-tests.py", line 1680, in do_test_wait_online_dns
[   98.561222] systemd-networkd-tests.py[691]:     self.fail(
[   98.561760] systemd-networkd-tests.py[691]:     ~~~~~~~~~^
[   98.562273] systemd-networkd-tests.py[691]:         f'Command timed out:\n{e.output}'
[   98.563223] systemd-networkd-tests.py[691]:         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[   98.563939] systemd-networkd-tests.py[691]:     )
[   98.565661] systemd-networkd-tests.py[691]:     ^
[   98.566276] systemd-networkd-tests.py[691]: AssertionError: Command timed out:
[   98.566815] systemd-networkd-tests.py[691]: Found link lo(1)
[   98.567342] systemd-networkd-tests.py[691]: Found link gre0(5)
[   98.567937] systemd-networkd-tests.py[691]: Found link gretap0(6)
[   98.568335] systemd-networkd-tests.py[691]: Found link erspan0(7)
[   98.568862] systemd-networkd-tests.py[691]: Found link veth-peer(24)
[   98.569808] systemd-networkd-tests.py[691]: Found link veth99(25)
[   98.570069] systemd-networkd-tests.py[691]: varlink: Setting state idle-client
[   98.570320] systemd-networkd-tests.py[691]: varlink: Sending message: {"method":"io.systemd.Resolve.Monitor.SubscribeDNSConfiguration","parameters":{"allowInteractiveAuthentication":false},"more":true}
[   98.571360] systemd-networkd-tests.py[691]: varlink: Changing state idle-client → awaiting-reply-more
[   98.571583] systemd-networkd-tests.py[691]: veth99: link is being processed by networkd: setup state is configuring.
[   98.571942] systemd-networkd-tests.py[691]: varlink: Received message: {"parameters":{"configuration":[{"currentServer":{"address":[192,168,5,1],"family":2,"port":53,"accessible":true},"servers":[{"address":[192,168,5,1],"family":2,"port":53,"accessible":true}]},{"ifname":"erspan0","ifindex":7,"defaultRoute":false},{"ifname":"gretap0","ifindex":6,"defaultRoute":false},{"ifname":"lo","ifindex":1,"defaultRoute":false},{"ifname":"veth-peer","ifindex":24,"defaultRoute":false},{"ifname":"gre0","ifindex":5,"defaultRoute":false},{"ifname":"veth99","ifindex":25,"defaultRoute":false}]},"continues":true}
[   98.572421] systemd-networkd-tests.py[691]: varlink: Changing state awaiting-reply-more → processing-reply
[   98.572728] systemd-networkd-tests.py[691]: veth99: link is being processed by networkd: setup state is configuring.
[   98.572917] systemd-networkd-tests.py[691]: varlink: Changing state processing-reply → awaiting-reply-more
[   98.573826] systemd-networkd-tests.py[691]: veth99: link is being processed by networkd: setup state is configuring.
[   98.574218] systemd-networkd-tests.py[691]: veth99: link is being processed by networkd: setup state is configuring.
[   98.574457] systemd-networkd-tests.py[691]: veth99: link is configured by networkd and online.
[   98.574745] systemd-networkd-tests.py[691]: varlink: Changing state awaiting-reply-more → disconnected
[   98.574951] systemd-networkd-tests.py[691]: =================================================================
[   98.575850] systemd-networkd-tests.py[691]: ==1299==ERROR: LeakSanitizer: detected memory leaks
[   98.576255] systemd-networkd-tests.py[691]: Direct leak of 5 byte(s) in 1 object(s) allocated from:
[   98.576530] systemd-networkd-tests.py[691]:     #0 0x7f4542ed4d53  (/usr/lib/clang/19/lib/x86_64-redhat-linux-gnu/libclang_rt.asan.so+0xd4d53) (BuildId: 1afad417992f0eb55159b853d8ca03bff039a365)
[   98.577619] systemd-networkd-tests.py[691]:     #1 0x7f45428e2a84  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xae2a84) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.578075] systemd-networkd-tests.py[691]:     #2 0x7f4542914e75  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb14e75) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.579288] systemd-networkd-tests.py[691]:     #3 0x55d9da6d697e  (/usr/lib/systemd/systemd-networkd-wait-online+0xa97e) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
[   98.580138] systemd-networkd-tests.py[691]:     #4 0x55d9da6d705e  (/usr/lib/systemd/systemd-networkd-wait-online+0xb05e) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
[   98.580554] systemd-networkd-tests.py[691]:     #5 0x7f4542914e75  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb14e75) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.581752] systemd-networkd-tests.py[691]:     #6 0x55d9da6d5ec0  (/usr/lib/systemd/systemd-networkd-wait-online+0x9ec0) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
[   98.581992] systemd-networkd-tests.py[691]:     #7 0x55d9da6e251f  (/usr/lib/systemd/systemd-networkd-wait-online+0x1651f) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
[   98.582993] systemd-networkd-tests.py[691]:     #8 0x7f4542922987  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb22987) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.583303] systemd-networkd-tests.py[691]:     #9 0x7f454293fc09  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb3fc09) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.583808] systemd-networkd-tests.py[691]:     #10 0x7f454279c151  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99c151) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.584065] systemd-networkd-tests.py[691]:     #11 0x7f454279a500  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99a500) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.592601] systemd-networkd-tests.py[691]:     #12 0x7f454279e698  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99e698) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.595089] systemd-networkd-tests.py[691]:     #13 0x7f454279ef0d  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99ef0d) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.596302] systemd-networkd-tests.py[691]:     #14 0x55d9da6e3b[46](https://github.com/systemd/systemd/actions/runs/13002269118/job/36263024103?pr=34640#step:12:47)  (/usr/lib/systemd/systemd-networkd-wait-online+0x17b46) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
[   98.597675] systemd-networkd-tests.py[691]:     #15 0x7f4541c105f4  (/lib64/libc.so.6+0x35f4) (BuildId: be08f175fb99297415ecc80c9ee7a33b56cbd355)
[   98.598375] systemd-networkd-tests.py[691]:     #16 0x7f4541c106a7  (/lib64/libc.so.6+0x36a7) (BuildId: be08f175fb99297415ecc80c9ee7a33b56cbd355)
[   98.598922] systemd-networkd-tests.py[691]:     #17 0x55d9da6d59b4  (/usr/lib/systemd/systemd-networkd-wait-online+0x99b4) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
[   98.599437] systemd-networkd-tests.py[691]: Direct leak of 5 byte(s) in 1 object(s) allocated from:
[   98.600881] systemd-networkd-tests.py[691]:     #0 0x7f4542ed4d53  (/usr/lib/clang/19/lib/x86_64-redhat-linux-gnu/libclang_rt.asan.so+0xd4d53) (BuildId: 1afad417992f0eb55159b853d8ca03bff039a365)
[   98.601267] systemd-networkd-tests.py[691]:     #1 0x7f45428e2a84  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xae2a84) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.601624] systemd-networkd-tests.py[691]:     #2 0x7f4542914e75  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb14e75) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.602109] systemd-networkd-tests.py[691]:     #3 0x55d9da6d697e  (/usr/lib/systemd/systemd-networkd-wait-online+0xa97e) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
[   98.602996] systemd-networkd-tests.py[691]:     #4 0x7f4542914e75  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb14e75) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.603412] systemd-networkd-tests.py[691]:     #5 0x55d9da6d5ec0  (/usr/lib/systemd/systemd-networkd-wait-online+0x9ec0) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
[   98.606660] systemd-networkd-tests.py[691]:     #6 0x55d9da6e251f  (/usr/lib/systemd/systemd-networkd-wait-online+0x1651f) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
[   98.607090] systemd-networkd-tests.py[691]:     #7 0x7f4542922987  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb22987) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.607459] systemd-networkd-tests.py[691]:     #8 0x7f454293fc09  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb3fc09) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.607936] systemd-networkd-tests.py[691]:     #9 0x7f454279c151  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99c151) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.608363] systemd-networkd-tests.py[691]:     #10 0x7f454279a500  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99a500) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.608861] systemd-networkd-tests.py[691]:     #11 0x7f454279e698  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99e698) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.609252] systemd-networkd-tests.py[691]:     #12 0x7f454279ef0d  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99ef0d) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
[   98.609686] systemd-networkd-tests.py[691]:     #13 0x55d9da6e3b46  (/usr/lib/systemd/systemd-networkd-wait-online+0x17b46) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
[   98.610089] systemd-networkd-tests.py[691]:     #14 0x7f4541c105f4  (/lib64/libc.so.6+0x35f4) (BuildId: be08f175fb99297415ecc80c9ee7a33b56cbd355)
[   98.6104[48](https://github.com/systemd/systemd/actions/runs/13002269118/job/36263024103?pr=34640#step:12:49)] systemd-networkd-tests.py[691]:     #15 0x7f4541c106a7  (/lib64/libc.so.6+0x36a7) (BuildId: be08f175fb99297415ecc80c9ee7a33b56cbd355)
[   98.610952] systemd-networkd-tests.py[691]:     #16 0x55d9da6d59b4  (/usr/lib/systemd/systemd-networkd-wait-online+0x99b4) (BuildId: 5f84a9682ff2e9c[50](https://github.com/systemd/systemd/actions/runs/13002269118/job/36263024103?pr=34640#step:12:51)2fbe8c5761f118a9b6860f5)
[   98.611442] systemd-networkd-tests.py[691]: SUMMARY: AddressSanitizer: 10 byte(s) leaked in 2 allocation(s).
[   98.611990] systemd-networkd-tests.py[691]: ----------------------------------------------------------------------
[   98.612383] systemd-networkd-tests.py[691]: Ran 5 tests in 70.632s
[   98.612862] systemd-networkd-tests.py[691]: FAILED (failures=2)
```

enr0n commented:
> This looks related @enr0n can you double check?
> 
> ```
> [   98.549876] systemd-networkd-tests.py[691]:   File "/usr/lib64/python3.13/subprocess.py", line 508, in check_returncode
> [   98.551901] systemd-networkd-tests.py[691]:     raise CalledProcessError(self.returncode, self.args, self.stdout,
> [   98.5529[45](https://github.com/systemd/systemd/actions/runs/13002269118/job/36263024103?pr=34640#step:12:46)] systemd-networkd-tests.py[691]:                              self.stderr)
> [   98.553440] systemd-networkd-tests.py[691]: subprocess.CalledProcessError: Command '['/usr/lib/systemd/systemd-networkd-wait-online', '--timeout=20s', '--interface=veth99:routable', '--ignore=sit0', '--ignore=ifb1', '--ignore=gre0', '--ignore=ifb0', '--ignore=lo', '--ignore=ip6tnl0', '--ignore=ip_vti0', '--ignore=ip6_vti0', '--ignore=teql0', '--ignore=ip6gre0', '--ignore=tunl0', '--ignore=gretap0', '--ignore=erspan0', '--dns']' returned non-zero exit status 1.
> [   98.553822] systemd-networkd-tests.py[691]: During handling of the above exception, another exception occurred:
> [   98.554372] systemd-networkd-tests.py[691]: Traceback (most recent call last):
> [   98.554753] systemd-networkd-tests.py[691]:   File "/usr/lib/systemd/tests/testdata/test-network/systemd-networkd-tests.py", line 1696, in test_wait_online_dns_global
> [   98.555136] systemd-networkd-tests.py[691]:     self.do_test_wait_online_dns(
> [   98.555498] systemd-networkd-tests.py[691]:     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
> [   98.556420] systemd-networkd-tests.py[691]:         global_dns='192.168.5.1',
> [   98.557201] systemd-networkd-tests.py[691]:         ^^^^^^^^^^^^^^^^^^^^^^^^^
> [   98.557664] systemd-networkd-tests.py[691]:     ...<3 lines>...
> [   98.558110] systemd-networkd-tests.py[691]:         )
> [   98.558487] systemd-networkd-tests.py[691]:         ^
> [   98.558915] systemd-networkd-tests.py[691]:     )
> [   98.560310] systemd-networkd-tests.py[691]:     ^
> [   98.560550] systemd-networkd-tests.py[691]:   File "/usr/lib/systemd/tests/testdata/test-network/systemd-networkd-tests.py", line 1680, in do_test_wait_online_dns
> [   98.561222] systemd-networkd-tests.py[691]:     self.fail(
> [   98.561760] systemd-networkd-tests.py[691]:     ~~~~~~~~~^
> [   98.562273] systemd-networkd-tests.py[691]:         f'Command timed out:\n{e.output}'
> [   98.563223] systemd-networkd-tests.py[691]:         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> [   98.563939] systemd-networkd-tests.py[691]:     )
> [   98.565661] systemd-networkd-tests.py[691]:     ^
> [   98.566276] systemd-networkd-tests.py[691]: AssertionError: Command timed out:
> [   98.566815] systemd-networkd-tests.py[691]: Found link lo(1)
> [   98.567342] systemd-networkd-tests.py[691]: Found link gre0(5)
> [   98.567937] systemd-networkd-tests.py[691]: Found link gretap0(6)
> [   98.568335] systemd-networkd-tests.py[691]: Found link erspan0(7)
> [   98.568862] systemd-networkd-tests.py[691]: Found link veth-peer(24)
> [   98.569808] systemd-networkd-tests.py[691]: Found link veth99(25)
> [   98.570069] systemd-networkd-tests.py[691]: varlink: Setting state idle-client
> [   98.570320] systemd-networkd-tests.py[691]: varlink: Sending message: {"method":"io.systemd.Resolve.Monitor.SubscribeDNSConfiguration","parameters":{"allowInteractiveAuthentication":false},"more":true}
> [   98.571360] systemd-networkd-tests.py[691]: varlink: Changing state idle-client → awaiting-reply-more
> [   98.571583] systemd-networkd-tests.py[691]: veth99: link is being processed by networkd: setup state is configuring.
> [   98.571942] systemd-networkd-tests.py[691]: varlink: Received message: {"parameters":{"configuration":[{"currentServer":{"address":[192,168,5,1],"family":2,"port":53,"accessible":true},"servers":[{"address":[192,168,5,1],"family":2,"port":53,"accessible":true}]},{"ifname":"erspan0","ifindex":7,"defaultRoute":false},{"ifname":"gretap0","ifindex":6,"defaultRoute":false},{"ifname":"lo","ifindex":1,"defaultRoute":false},{"ifname":"veth-peer","ifindex":24,"defaultRoute":false},{"ifname":"gre0","ifindex":5,"defaultRoute":false},{"ifname":"veth99","ifindex":25,"defaultRoute":false}]},"continues":true}
> [   98.572421] systemd-networkd-tests.py[691]: varlink: Changing state awaiting-reply-more → processing-reply
> [   98.572728] systemd-networkd-tests.py[691]: veth99: link is being processed by networkd: setup state is configuring.
> [   98.572917] systemd-networkd-tests.py[691]: varlink: Changing state processing-reply → awaiting-reply-more
> [   98.573826] systemd-networkd-tests.py[691]: veth99: link is being processed by networkd: setup state is configuring.
> [   98.574218] systemd-networkd-tests.py[691]: veth99: link is being processed by networkd: setup state is configuring.
> [   98.574457] systemd-networkd-tests.py[691]: veth99: link is configured by networkd and online.
> [   98.574745] systemd-networkd-tests.py[691]: varlink: Changing state awaiting-reply-more → disconnected
> [   98.574951] systemd-networkd-tests.py[691]: =================================================================
> [   98.575850] systemd-networkd-tests.py[691]: ==1299==ERROR: LeakSanitizer: detected memory leaks
> [   98.576255] systemd-networkd-tests.py[691]: Direct leak of 5 byte(s) in 1 object(s) allocated from:
> [   98.576530] systemd-networkd-tests.py[691]:     #0 0x7f4542ed4d53  (/usr/lib/clang/19/lib/x86_64-redhat-linux-gnu/libclang_rt.asan.so+0xd4d53) (BuildId: 1afad417992f0eb55159b853d8ca03bff039a365)
> [   98.577619] systemd-networkd-tests.py[691]:     #1 0x7f45428e2a84  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xae2a84) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.578075] systemd-networkd-tests.py[691]:     #2 0x7f4542914e75  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb14e75) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.579288] systemd-networkd-tests.py[691]:     #3 0x55d9da6d697e  (/usr/lib/systemd/systemd-networkd-wait-online+0xa97e) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
> [   98.580138] systemd-networkd-tests.py[691]:     #4 0x55d9da6d705e  (/usr/lib/systemd/systemd-networkd-wait-online+0xb05e) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
> [   98.580554] systemd-networkd-tests.py[691]:     #5 0x7f4542914e75  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb14e75) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.581752] systemd-networkd-tests.py[691]:     #6 0x55d9da6d5ec0  (/usr/lib/systemd/systemd-networkd-wait-online+0x9ec0) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
> [   98.581992] systemd-networkd-tests.py[691]:     #7 0x55d9da6e251f  (/usr/lib/systemd/systemd-networkd-wait-online+0x1651f) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
> [   98.582993] systemd-networkd-tests.py[691]:     #8 0x7f4542922987  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb22987) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.583303] systemd-networkd-tests.py[691]:     #9 0x7f454293fc09  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb3fc09) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.583808] systemd-networkd-tests.py[691]:     #10 0x7f454279c151  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99c151) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.584065] systemd-networkd-tests.py[691]:     #11 0x7f454279a500  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99a500) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.592601] systemd-networkd-tests.py[691]:     #12 0x7f454279e698  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99e698) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.595089] systemd-networkd-tests.py[691]:     #13 0x7f454279ef0d  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99ef0d) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.596302] systemd-networkd-tests.py[691]:     #14 0x55d9da6e3b[46](https://github.com/systemd/systemd/actions/runs/13002269118/job/36263024103?pr=34640#step:12:47)  (/usr/lib/systemd/systemd-networkd-wait-online+0x17b46) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
> [   98.597675] systemd-networkd-tests.py[691]:     #15 0x7f4541c105f4  (/lib64/libc.so.6+0x35f4) (BuildId: be08f175fb99297415ecc80c9ee7a33b56cbd355)
> [   98.598375] systemd-networkd-tests.py[691]:     #16 0x7f4541c106a7  (/lib64/libc.so.6+0x36a7) (BuildId: be08f175fb99297415ecc80c9ee7a33b56cbd355)
> [   98.598922] systemd-networkd-tests.py[691]:     #17 0x55d9da6d59b4  (/usr/lib/systemd/systemd-networkd-wait-online+0x99b4) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
> [   98.599437] systemd-networkd-tests.py[691]: Direct leak of 5 byte(s) in 1 object(s) allocated from:
> [   98.600881] systemd-networkd-tests.py[691]:     #0 0x7f4542ed4d53  (/usr/lib/clang/19/lib/x86_64-redhat-linux-gnu/libclang_rt.asan.so+0xd4d53) (BuildId: 1afad417992f0eb55159b853d8ca03bff039a365)
> [   98.601267] systemd-networkd-tests.py[691]:     #1 0x7f45428e2a84  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xae2a84) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.601624] systemd-networkd-tests.py[691]:     #2 0x7f4542914e75  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb14e75) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.602109] systemd-networkd-tests.py[691]:     #3 0x55d9da6d697e  (/usr/lib/systemd/systemd-networkd-wait-online+0xa97e) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
> [   98.602996] systemd-networkd-tests.py[691]:     #4 0x7f4542914e75  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb14e75) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.603412] systemd-networkd-tests.py[691]:     #5 0x55d9da6d5ec0  (/usr/lib/systemd/systemd-networkd-wait-online+0x9ec0) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
> [   98.606660] systemd-networkd-tests.py[691]:     #6 0x55d9da6e251f  (/usr/lib/systemd/systemd-networkd-wait-online+0x1651f) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
> [   98.607090] systemd-networkd-tests.py[691]:     #7 0x7f4542922987  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb22987) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.607459] systemd-networkd-tests.py[691]:     #8 0x7f454293fc09  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0xb3fc09) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.607936] systemd-networkd-tests.py[691]:     #9 0x7f454279c151  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99c151) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.608363] systemd-networkd-tests.py[691]:     #10 0x7f454279a500  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99a500) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.608861] systemd-networkd-tests.py[691]:     #11 0x7f454279e698  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99e698) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.609252] systemd-networkd-tests.py[691]:     #12 0x7f454279ef0d  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x99ef0d) (BuildId: e538e087c19b2620eedc017e63e73cf668e4f137)
> [   98.609686] systemd-networkd-tests.py[691]:     #13 0x55d9da6e3b46  (/usr/lib/systemd/systemd-networkd-wait-online+0x17b46) (BuildId: 5f84a9682ff2e9c502fbe8c5761f118a9b6860f5)
> [   98.610089] systemd-networkd-tests.py[691]:     #14 0x7f4541c105f4  (/lib64/libc.so.6+0x35f4) (BuildId: be08f175fb99297415ecc80c9ee7a33b56cbd355)
> [   98.6104[48](https://github.com/systemd/systemd/actions/runs/13002269118/job/36263024103?pr=34640#step:12:49)] systemd-networkd-tests.py[691]:     #15 0x7f4541c106a7  (/lib64/libc.so.6+0x36a7) (BuildId: be08f175fb99297415ecc80c9ee7a33b56cbd355)
> [   98.610952] systemd-networkd-tests.py[691]:     #16 0x55d9da6d59b4  (/usr/lib/systemd/systemd-networkd-wait-online+0x99b4) (BuildId: 5f84a9682ff2e9c[50](https://github.com/systemd/systemd/actions/runs/13002269118/job/36263024103?pr=34640#step:12:51)2fbe8c5761f118a9b6860f5)
> [   98.611442] systemd-networkd-tests.py[691]: SUMMARY: AddressSanitizer: 10 byte(s) leaked in 2 allocation(s).
> [   98.611990] systemd-networkd-tests.py[691]: ----------------------------------------------------------------------
> [   98.612383] systemd-networkd-tests.py[691]: Ran 5 tests in 70.632s
> [   98.612862] systemd-networkd-tests.py[691]: FAILED (failures=2)
> ```

I was missing an `iovec_done()` in `dns_server_free()`.

enr0n commented:
Okay, I believe the remaining failures are unrelated now.

poettering commented on src/resolve/resolved-varlink.c:
```diff
@@ -1354,6 +1364,45 @@ static int vl_method_reset_statistics(sd_varlink *link, sd_json_variant *paramet
+        if (set_isempty(m->varlink_dns_configuration_subscription))
+                manager_stop_dns_configuration_monitor(m);
+
+
```
spurious double empty line

enr0n replied:
> spurious double empty line

https://github.com/systemd/systemd/pull/36216
